baseplate/experiments/__init__.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import json
+import logging
+
+from collections import namedtuple
+
+from .providers import parse_experiment
+from .._compat import iteritems
+from ..context import ContextFactory
+from ..events import Event, EventTooLargeError, EventQueueFullError
+from ..file_watcher import FileWatcher, WatchedFileNotAvailableError
+
+
+logger = logging.getLogger(__name__)
+
+
+class User(object):
+
+    def __init__(self, id, name, created, flags=None):
+        self.id = id
+        self.name = name
+        self.created = created
+        self.flags = flags or set()
+
+    @property
+    def logged_in(self):
+        return self.name is not None
+
+
+Content = namedtuple("Content", ["id", "type"])
+
+
+class ExperimentsContextFactory(ContextFactory):
+
+    def __init__(self, path, event_queue):
+        self._filewatcher = FileWatcher(path, json.load)
+        self._event_queue = event_queue
+
+    def make_object_for_context(self, name, server_span):
+        return Experiments(self._filewatcher, self._event_queue)
+
+
+class Experiments(object):
+
+    def __init__(self, config_watcher, event_queue=None, session=None):
+        self._config_watcher = config_watcher
+        self._event_queue = event_queue
+        self._cache = {}
+        self._session = session
+
+    def _get_config(self, name):
+        try:
+            config_data = self._config_watcher.get_data()
+            return config_data[name]
+        except WatchedFileNotAvailableError:
+            logger.warning("Experiment config file not found")
+            return
+        except KeyError:
+            logger.warning(
+                "Experiment <%r> not found in experiment config",
+                name,
+            )
+            return
+        except TypeError:
+            logger.warning("Could not load experiment config: %r", name)
+            return
+
+    def variant(self, name, bucketing_event_override=None,
+                extra_event_params=None, **kwargs):
+        cache_key = "%s:%s" % (name, str(sorted(iteritems(kwargs))))
```

Comment:
I'll note, I'm not enthusiastic about this cache key scheme.  I could just cache based on the value used for bucketing and the experiment name, which is what r2 does.  My only concern with that is r2 pulls all the targeting information out of the aether (`g` & `c`) but with this you can ask for a variant for the same user but with different targeting parameters and just caching by bucketing value would not capture that.